File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/0095b2554ebb25123ac6/repair-v1/job_byox_repair_v1_g1_798ae60980096ac6d6f48e11dde7ba6c/attempt-001/sealed/reference/vm.go
package pebble

import "math"

func Run(code Bytecode) ([]int64, error) {
	if err := ValidateBytecode(code); err != nil {
		return nil, err
	}
	stack := make([]int64, 0)
	locals := make(map[int64]int64)
	output := make([]int64, 0)
	for _, instruction := range code.Instructions {
		switch instruction.Op {
		case OpPush:
			stack = append(stack, instruction.Operand)
		case OpLoad:
			stack = append(stack, locals[instruction.Operand])
		case OpStore:
			value := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			locals[instruction.Operand] = value
		case OpAdd, OpSub, OpMul, OpDiv:
			right := stack[len(stack)-1]
			left := stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			value, code := calculate(instruction.Op, left, right)
			if code != "" {
				message := "integer arithmetic overflow"
				if code == CodeDivisionByZero {
					message = "division by zero"
				}
				return nil, languageError(StageRun, code, instruction.Span.Start, message)
			}
			stack = append(stack, value)
		case OpPrint:
			value := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			output = append(output, value)
		case OpPop:
			stack = stack[:len(stack)-1]
		case OpHalt:
			return output, nil
		}
	}
	return nil, languageError(StageValidate, CodeInvalidBytecode, Position{Line: 1, Column: 1}, "validated stream did not halt")
}

func calculate(op OpCode, left, right int64) (int64, string) {
	switch op {
	case OpAdd:
		if right > 0 && left > math.MaxInt64-right || right < 0 && left < math.MinInt64-right {
			return 0, CodeIntegerOverflow
		}
		return left + right, ""
	case OpSub:
		if right > 0 && left < math.MinInt64+right || right < 0 && left > math.MaxInt64+right {
			return 0, CodeIntegerOverflow
		}
		return left - right, ""
	case OpMul:
		if left == 0 || right == 0 {
			return 0, ""
		}
		if left == -1 && right == math.MinInt64 || right == -1 && left == math.MinInt64 {
			return 0, CodeIntegerOverflow
		}
		value := left * right
		if value/right != left {
			return 0, CodeIntegerOverflow
		}
		return value, ""
	case OpDiv:
		if right == 0 {
			return 0, CodeDivisionByZero
		}
		if left == math.MinInt64 && right == -1 {
			return 0, CodeIntegerOverflow
		}
		return left / right, ""
	default:
		return 0, CodeIntegerOverflow
	}
}

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/2af0535ee2e2751f577b/baseline-c44cc5b11936d582697d/repair-v1/job_byox_repair_s2_v1_g1_88f6c80cd38ab51517a185f39563426c/attempt-001/sealed/reference/src/vm.c
#include "sprig.h"

#include <inttypes.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>

static int runtime_error(Diagnostic *diagnostic,
                         const Instruction *instruction,
                         const char *message) {
    diagnostic->line = instruction == NULL ? 1u : instruction->line;
    diagnostic->column = instruction == NULL ? 1u : instruction->column;
    (void)snprintf(diagnostic->message, sizeof(diagnostic->message),
                   "%s", message);
    return 0;
}

static int checked_add(int64_t left, int64_t right, int64_t *result) {
    if ((right > 0 && left > INT64_MAX - right) ||
        (right < 0 && left < INT64_MIN - right)) {
        return 0;
    }
    *result = left + right;
    return 1;
}

static int checked_sub(int64_t left, int64_t right, int64_t *result) {
    if ((right > 0 && left < INT64_MIN + right) ||
        (right < 0 && left > INT64_MAX + right)) {
        return 0;
    }
    *result = left - right;
    return 1;
}

static int checked_mul(int64_t left, int64_t right, int64_t *result) {
    if (left == 0 || right == 0) {
        *result = 0;
        return 1;
    }
    if (left > 0) {
        if ((right > 0 && left > INT64_MAX / right) ||
            (right < 0 && right < INT64_MIN / left)) {
            return 0;
        }
    } else {
        if ((right > 0 && left < INT64_MIN / right) ||
            (right < 0 && left < INT64_MAX / right)) {
            return 0;
        }
    }
    *result = left * right;
    return 1;
}

static int checked_binary(OpCode opcode, int64_t left, int64_t right,
                          int64_t *result, const char **error) {
    switch (opcode) {
        case OP_ADD:
            if (!checked_add(left, right, result)) {
                *error = "integer overflow in addition";
                return 0;
            }
            return 1;
        case OP_SUB:
            if (!checked_sub(left, right, result)) {
                *error = "integer overflow in subtraction";
                return 0;
            }
            return 1;
        case OP_MUL:
            if (!checked_mul(left, right, result)) {
                *error = "integer overflow in multiplication";
                return 0;
            }
            return 1;
        case OP_DIV:
            if (right == 0) {
                *error = "division by zero";
                return 0;
            }
            if (left == INT64_MIN && right == -1) {
                *error = "integer overflow in division";
                return 0;
            }
            *result = left / right;
            return 1;
        default:
            *error = "invalid binary opcode";
            return 0;
    }
}

int vm_execute(const Program *program, FILE *output, Diagnostic *diagnostic) {
    int64_t stack[SPRIG_MAX_STACK];
    int64_t variables[SPRIG_MAX_VARIABLES];
    unsigned char initialized[SPRIG_MAX_VARIABLES];
    size_t stack_count = 0u;
    size_t pc;

    memset(diagnostic, 0, sizeof(*diagnostic));
    memset(variables, 0, sizeof(variables));
    memset(initialized, 0, sizeof(initialized));
    if (program->count == 0u ||
        program->count > SPRIG_MAX_INSTRUCTIONS ||
        program->variable_count > SPRIG_MAX_VARIABLES) {
        return runtime_error(diagnostic, NULL, "invalid program metadata");
    }

    for (pc = 0u; pc < program->count; pc++) {
        const Instruction *instruction = &program->code[pc];
        int64_t left;
        int64_t right;
        int64_t result;
        const char *arithmetic_error = NULL;
        size_t slot;

        switch (instruction->opcode) {
            case OP_CONST:
                if (stack_count >= SPRIG_MAX_STACK) {
                    return runtime_error(diagnostic, instruction,
                                         "value stack overflow");
                }
                stack[stack_count++] = instruction->operand;
                break;
            case OP_LOAD:
                if (instruction->operand < 0 ||
                    (uint64_t)instruction->operand >= program->variable_count) {
                    return runtime_error(diagnostic, instruction,
                                         "invalid variable slot in LOAD");
                }
                slot = (size_t)instruction->operand;
                if (!initialized[slot]) {
                    return runtime_error(diagnostic, instruction,
                                         "LOAD from uninitialized slot");
                }
                if (stack_count >= SPRIG_MAX_STACK) {
                    return runtime_error(diagnostic, instruction,
                                         "value stack overflow");
                }
                stack[stack_count++] = variables[slot];
                break;
            case OP_STORE:
                if (instruction->operand < 0 ||
                    (uint64_t)instruction->operand >= program->variable_count) {
                    return runtime_error(diagnostic, instruction,
                                         "invalid variable slot in STORE");
                }
                if (stack_count == 0u) {
                    return runtime_error(diagnostic, instruction,
                                         "value stack underflow in STORE");
                }
                slot = (size_t)instruction->operand;
                if (initialized[slot]) {
                    return runtime_error(diagnostic, instruction,
                                         "STORE to initialized slot");
                }
                variables[slot] = stack[--stack_count];
                initialized[slot] = 1u;
                break;
            case OP_ADD:
            case OP_SUB:
            case OP_MUL:
            case OP_DIV:
                if (stack_count < 2u) {
                    return runtime_error(diagnostic, instruction,
                                         "value stack underflow in arithmetic");
                }
                right = stack[--stack_count];
                left = stack[--stack_count];
                if (!checked_binary(instruction->opcode, left, right,
                                    &result, &arithmetic_error)) {
                    return runtime_error(diagnostic, instruction,
                                         arithmetic_error);
                }
                stack[stack_count++] = result;
                break;
            case OP_NEG:
                if (stack_count == 0u) {
                    return runtime_error(diagnostic, instruction,
                                         "value stack underflow in NEG");
                }
                if (stack[stack_count - 1u] == INT64_MIN) {
                    return runtime_error(diagnostic, instruction,
                                         "integer overflow in negation");
                }
                stack[stack_count - 1u] = -stack[stack_count - 1u];
                break;
            case OP_PRINT:
                if (stack_count == 0u) {
                    return runtime_error(diagnostic, instruction,
                                         "value stack underflow in PRINT");
                }
                if (fprintf(output, "%" PRId64 "\n",
                            stack[--stack_count]) < 0) {
                    return runtime_error(diagnostic, instruction,
                                         "failed to write program output");
                }
                break;
            case OP_HALT:
                if (pc + 1u != program->count) {
                    return runtime_error(diagnostic, instruction,
                                         "instructions follow HALT");
                }
                if (stack_count != 0u) {
                    return runtime_error(diagnostic, instruction,
                                         "HALT with non-empty value stack");
                }
                if (fflush(output) != 0) {
                    return runtime_error(diagnostic, instruction,
                                         "failed to flush program output");
                }
                return 1;
            default:
                return runtime_error(diagnostic, instruction,
                                     "unknown opcode");
        }
    }
    return runtime_error(diagnostic, &program->code[program->count - 1u],
                         "program has no HALT instruction");
}
